crates/re_arrow_store/src/store_event.rs

```diff
@@ -0,0 +1,494 @@
+///
+/// Methods that mutate the [`DataStore`], such as [`DataStore::insert_row`] and [`DataStore::gc`],
+/// return [`StoreEvent`]s that describe the changes.
+///
+/// Refer to field-level documentation for more details and check out [`StoreDiff`] for a precise
+/// definition of what an event involves.
+#[derive(Debug, Clone, PartialEq)]
+pub struct StoreEvent {
+    /// Which [`DataStore`] sent this event?
+    pub store_id: StoreId,
+
+    /// What was the store's generation when it sent that event?
+    pub store_generation: StoreGeneration,
+
+    /// Monotonically increasing ID of the event.
+    ///
+    /// This is on a per-store basis.
+    ///
+    /// When handling a [`StoreEvent`], if this is the first time you process this [`StoreId`] and
+    /// the associated `event_id` is not `1`, it means you registered late and missed some updates.
+    pub event_id: u64,
+
+    /// What actually changed?
+    ///
+    /// Refer to [`StoreDiff`] for more information.
+    pub diff: StoreDiff,
+}
+
+impl std::ops::Deref for StoreEvent {
+    type Target = StoreDiff;
+
+    fn deref(&self) -> &Self::Target {
+        &self.diff
+    }
+}
+
+/// Is it an addition or a deletion?
+///
+/// Reminder: ⚠ Do not confuse _a deletion_ and _a clear_ ⚠.
+///
+/// A deletion is the result of a row being completely removed from the store as part of the
+/// garbage collection process.
+///
+/// A clear, on the other hand, is the act of logging an empty [`re_types_core::ComponentBatch`],
+/// either directly using the logging APIs, or indirectly through the use of a
+/// [`re_types_core::archetypes::Clear`] archetype.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum StoreDiffKind {
+    Addition,
+    Deletion,
+}
+
+impl StoreDiffKind {
+    #[inline]
+    pub fn delta(&self) -> i64 {
+        match self {
+            StoreDiffKind::Addition => 1,
+            StoreDiffKind::Deletion => -1,
+        }
+    }
+}
+
+/// Describes an atomic change in the Rerun [`DataStore`]: a row has been added or deleted.
+///
+/// From a query model standpoint, the [`DataStore`] _always_ operates one row at a time:
+/// - The contents of a row (i.e. its columns) are immutable past insertion, by virtue of
+///   [`RowId`]s being unique and non-reusable.
+/// - Similarly, garbage collection always removes _all the data_ associated with a row in one go:
+///   there cannot be orphaned columns. When a row is gone, all data associated with it is gone too.
+///
+/// Refer to field-level documentation for more information.
+#[derive(Debug, Clone, PartialEq)]
+pub struct StoreDiff {
+    /// Addition or deletion?
+    ///
+    /// The store's internals are opaque and don't necessarily reflect the query model (e.g. there
+    /// might be data in the store that cannot by reached by any query).
+    ///
+    /// A [`StoreDiff`] answers a logical question: "does there exist a query path which can return
+    /// data from that row?".
```

Comment:
makes sense, but what has the reachability of rows to do with whether this is an addition or deletion?

Reply:
That's the point I'm trying to make: an event of kind deletion only tells you that, from this point on, no query can return data from that row.

_But_ that doesn't mean that the data is actually gone, i.e. don't make assumptions of e.g. the size in bytes of the store based on these events. They are in "query-model space" and are not an accurate representation of what happens in storage space.